src/bucket/BucketIndex.cpp

```diff
@@ -0,0 +1,339 @@
+    }
+    else
+    {
+        return indexEntry.first < key;
+    }
+}
+
+// std::upper_bound predicate. Returns true if key comes "before" and is not
+// contained within the indexEntry.
+// If key is too small for indexEntry bounds: return true
+// If key is contained within indexEntry bounds: return false
+// If key is too large for indexEntry bounds: return false
+template <class IndexEntryT>
+static bool
+upper_bound_pred(LedgerKey const& key, IndexEntryT const& indexEntry)
+{
+    if constexpr (std::is_same<IndexEntryT,
+                               BucketIndex::RangeIndex::value_type>::value)
+    {
+        return key < indexEntry.first.lowerBound;
+    }
+    else
+    {
+        return key < indexEntry.first;
+    }
+}
+
+std::unique_ptr<BucketIndex const>
+BucketIndex::createIndex(BucketManager const& bm,
+                         std::filesystem::path const& filename)
+{
+    ZoneScoped;
+    auto const& cfg = bm.getConfig();
+    releaseAssertOrThrow(cfg.EXPERIMENTAL_BUCKETLIST_DB);
+    releaseAssertOrThrow(!filename.empty());
+
+    auto pageSizeExp = cfg.EXPERIMENTAL_BUCKETLIST_DB_INDEX_PAGE_SIZE_EXPONENT;
+    releaseAssertOrThrow(pageSizeExp < 32);
+    auto pageSize = pageSizeExp == 0 ? 0 : 1UL << pageSizeExp;
+
+    // Conver to bytes
+    auto cutoff = cfg.EXPERIMENTAL_BUCKETLIST_DB_INDEX_CUTOFF * 1000000;
+    if (pageSize == 0 || fs::size(filename.string()) < cutoff)
+    {
+        CLOG_INFO(Bucket,
+                  "BucketIndex::createIndex() indexing individual keys in "
+                  "bucket {}",
+                  filename);
+        return std::unique_ptr<BucketIndexImpl<IndividualIndex> const>(
+            new BucketIndexImpl<IndividualIndex>(bm, filename, 0));
+    }
+    else
+    {
+        CLOG_INFO(Bucket,
+                  "BucketIndex::createIndex() indexing key range with "
+                  "page size "
+                  "{} in bucket {}",
+                  pageSize, filename);
+        return std::unique_ptr<BucketIndexImpl<RangeIndex> const>(
+            new BucketIndexImpl<RangeIndex>(bm, filename, pageSize));
+    }
+
+    return {};
+}
+
+template <class IndexT>
+std::optional<std::streamoff>
+BucketIndexImpl<IndexT>::lookup(LedgerKey const& k) const
+{
+    ZoneScoped;
+    return scan(begin(), k).first;
+}
+
+template <class IndexT>
+std::pair<std::optional<std::streamoff>, BucketIndex::Iterator>
+BucketIndexImpl<IndexT>::scan(Iterator start, LedgerKey const& k) const
+{
+    ZoneScoped;
+    ZoneValue(static_cast<int64_t>(mKeysToOffset.size()));
+
+    auto internalStart = std::get<typename IndexT::const_iterator>(start);
+    auto keyIter =
+        std::lower_bound(internalStart, mKeysToOffset.end(), k,
+                         lower_bound_pred<typename IndexT::value_type>);
+
+    // If the key is not in the bloom filter or in the lower bounded index
+    // entry, return nullopt
+    if ((mFilter && !mFilter->contains(std::hash<stellar::LedgerKey>()(k))) ||
```

Comment:
Bloom filters don't give false negatives, only false positives, right? Can you not check the filter before the binary search? (I tried it myself here and it does not seem to work, which suggests I don't understand something about how this works)

Reply:
I went back and forth on this one actually. Initially, I did check the bloom filter first and exit early before doing the `std::lower_bound` search. However, when you check the bloom filter first, the `scan` interface isn't quite right. The caller expects `scan` to return an iterator to `k` or to the first element greater than `k`. If we check the bloom filter before binary search, this interface is broken. We either just return the `start` iterator passed in as a parameter or some sort of null iterator. If we return a null iterator, the caller loses there place in the index, so follow up calls to `scan` will be less efficient. If we return `start`, everything works but we're not following the interface properly. For me, binary search first seems like the correct option. This way we can return the correct iterator that the caller expects, but still use the bloom filter to avoid expensive disk io. I think if we change the interface to return `start` if the key is not in the bloom filter, we are slightly more efficient. However this means the `scan` interface is different between `RangeIndex` and `IndividualIndex`, which is an implementation detail I'd rather not leak to the caller.

Reply:
Ok. I suspect it's a bit of a speed hit but I understand the rationale. It might be nice to put a comment in about this, because any future reader who knows "bloom filters are to save you work when it gets a negative hit" will consider hoisting it up.